design/src/rfcs/rfc0004_retry_behavior.md

```diff
@@ -0,0 +1,194 @@
+    let resp = sts.get_caller_identity().send().await?;
+    println!("your user id: {}", resp.user_id.unwrap_or_default());
+    Ok(())
+}
+```
+
+Retries can also be disabled by explicitly passing the `RetryConfig::NoRetries` enum variant to the `retry_config` builder method:
+
+```rust
+use aws_sdk_sts as sts;
+use aws_types::retry_config::RetryConfig;
+
+#[tokio::main]
+async fn main() -> Result<(), sts::Error> {
+    let config = aws_config::load_from_env().await;
+    let sts_config = sts::config::Config::from(&config).retry_config(RetryConfig::NoRetries).build();
+
+    let sts = sts::Client::new(&sts_config);
+    let resp = sts.get_caller_identity().send().await?;
+    println!("your user id: {}", resp.user_id.unwrap_or_default());
+    Ok(())
+}
+```
+
+Behind the scenes
+-----------------
+
+Currently, when users want to send a request, the following occurs:
+
+1. The user creates either a shared config or a service-specific config
+1. The user creates a fluent client for the service they want to interact with and passes the config they created. Internally, this creates an AWS client with a default retry policy
+1. The user calls an operation builder method on the client which constructs a request
+1. The user sends the request by awaiting the `send()` method
+1. The smithy client creates a new `Service` and attaches a copy of its retry policy
+1. The `Service` is `call`ed, sending out the request and retrying it according to the retry policy
+
+After this change, the process will work like this:
+
+1. The user creates either a shared config or a service-specific config
+    - If `AWS_MAX_ATTEMPTS` is set to zero, this is invalid. However, this will not error until a request is made
+    - If `AWS_MAX_ATTEMPTS` is 1, retries will be disabled
+    - If `AWS_MAX_ATTEMPTS` is greater than 1, retries will be attempted at most as many times as is specified
+    - If the user creates the config with the `.disable_retries` builder method, retries will be disabled
+    - If the user creates the config with the `retry_config` builder method, retry behavior will be set according to the `RetryConfig` they passed
+1. The user creates a fluent client for the service they want to interact with and passes the config they created
+    - Provider precedence will determine what retry behavior is actually set, working like how `Region` is set
+1. The user calls an operation builder method on the client which constructs a request
+1. The user sends the request by awaiting the `send()` method
+1. The smithy client creates a new `Service` and attaches a copy of its retry policy
+1. The `Service` is `call`ed, sending out the request and retrying it according to the retry policy
+
+These changes will be made in such a way that they enable us to add the "adaptive" retry behavior at a later date without introducing a breaking change.
+
+Changes checklist
+-----------------
+
+- [ ] Create new Kotlin decorator `RetryConfigDecorator`
```

Comment:
I did not realize we had non-generated clients, do you have an example of one I can look at?

Reply:
Oh, sorry, I was unclear. I meant clients that are generated with smithy-rs that are not for the AWS SDK. The SDK is one use of smithy-rs to generate clients, but there can be others since it's intended to be a generic Smithy Rust code generator. We don't have any examples in the repo currently.

Reply:
Got it, I'll write a note about putting in the `codegen` project